src/coprocessor/dag/batch/executors/selection_executor.rs

```diff
@@ -58,6 +58,101 @@ impl<Src: BatchExecutor> BatchSelectionExecutor<Src> {
+        // have error, which should be filtered according to predicate in this executor.
+        // So we actually don't care whether or not there are errors from src executor.
+
+        // TODO: Avoid allocation.
+        let mut src_logical_rows_copy = Vec::with_capacity(src_result.logical_rows.len());
+        let mut condition_index = 0;
+        while condition_index < self.conditions.len() && !src_result.logical_rows.is_empty() {
+            src_logical_rows_copy.clear();
+            src_logical_rows_copy.extend_from_slice(&src_result.logical_rows);
+
+            match self.conditions[condition_index].eval(
+                &mut self.context,
+                self.src.schema(),
+                &mut src_result.physical_columns,
+                &src_logical_rows_copy,
+                src_logical_rows_copy.len(),
+            )? {
+                RpnStackNode::Scalar { value, .. } => {
+                    update_logical_rows_by_scalar_value(
+                        &mut src_result.logical_rows,
+                        &mut self.context,
+                        value,
+                    )?;
+                }
+                RpnStackNode::Vector { value, .. } => {
+                    let eval_result_logical_rows = value.logical_rows();
+                    match_template_evaluable! {
+                        TT, match value.as_ref() {
+                            VectorValue::TT(eval_result) => {
+                                update_logical_rows_by_vector_value(
+                                    &mut src_result.logical_rows,
+                                    &mut self.context,
+                                    eval_result,
+                                    eval_result_logical_rows,
+                                )?;
+                            },
+                        }
+                    }
+                }
+            }
+
+            condition_index += 1;
+        }
+
+        Ok(())
+    }
+}
+
+fn update_logical_rows_by_scalar_value(
+    logical_rows: &mut Vec<usize>,
+    context: &mut EvalContext,
+    value: &ScalarValue,
+) -> Result<()> {
+    let b = value.as_mysql_bool(context)?;
+    if !b {
+        // No rows should be preserved
+        logical_rows.clear();
+    }
+    Ok(())
+}
+
+fn update_logical_rows_by_vector_value<T: AsMySQLBool>(
+    logical_rows: &mut Vec<usize>,
+    context: &mut EvalContext,
+    eval_result: &[Option<T>],
+    eval_result_logical_rows: &[usize],
+) -> Result<()> {
+    let mut err_result = Ok(());
+    let mut logical_index = 0;
+    logical_rows.retain(|_| {
+        // We don't care the physical index indicated by `logical_rows`, since what's in there
+        // does not affect the filtering. Instead, the eval result in corresponding logical index
+        // matters.
+
+        let eval_result_physical_index = eval_result_logical_rows[logical_index];
+        logical_index += 1;
```

Comment:
This is somewhat subtle. It doesn't seem a good idea that the predicate is a captured variable because  `retain` does not guarantee to call the predicate function in order...

Reply:
Indeed.. And interestingly servo's Vec ensures order: https://doc.servo.org/alloc/vec/struct.Vec.html#method.retain

Reply:
Don't worry! Retain order is now documented recently: https://github.com/rust-lang/rust/pull/60396